ably/ably-chat-js
# react: use stable references to event listeners in hooks

### Context

N/A

### Description

A number of our hooks allow users to pass in their own listeners to be able to monitor events such as new messages or occupancy updates, even if the hook exposes the latest data.

At present, we use the listener param in a useEffect to apply listener subscriptions. This causes listener cycling if the hook is called using an inline listener, as the reference changes every time. This behaviour is not desired, as it may cause the subscriber to miss events.

This change fixes the issue by introducing a new internal hook, useEventListenerRef. This hook creates a stable reference to the passed in listener between re-renders. So whilst the underlying listener may change, we avoid cycling subscriptions on every render and thus potentially missing messages.

We also replace all instances of listener use in the other hooks with calls to this new hook, so that the listener passed in is maintained.

### Checklist


### Testing Instructions (Optional)

On main, try the in-repo demo app using trace level logging. If you type something into the input, you'll see that the useRoomReactions listener unmounts and remounts.

On this branch, you'll see that no longer happens.
